### DISCUSSION TITLE
Extracting local primal solution at the current node

### DISCUSSION THREAD

**Discussion by cwfparsonson** [ds4dm/ecole#279] [Questions] [answered]
Hi,

In `pyscipopt`, it possible to get the objective value of the primal solution at the current node, similar to how you can do `node.getLowerbound()`? `node.getEstimate()` seems to give quite erratic results and I don't think it is related to the true primal bound objective value at the current node.

I think you can do `primal_obj_vals = [m.getSolObjVal(sol) for sol in m.getSols()]` to get a list of the primal objective values found so far, but I can't see how to relate this to the current local node.

Thanks

**Comment by AntoinePrv** [comment]
Hi @cwfparsonson 

In banch-and-bound there isn't necessarily a primal solution at every node. One may get a primal solution if the solution to the relaxation is integer feasible (not always the case). In modern solvers, primal solutions can also be obtained by running primal heuristics, but this is not done on every node, nor is it guaranteed to  give a solution.

**Reply by cwfparsonson** [reply]
Sorry, I meant where a feasible primal solution exists and has been generated by SCIP at the current node, is it possible to extract?

**Comment by gasse** [comment, marked as answer]
Hi @cwfparsonson ,

I think you should be able to do what you want by registering an event handler
https://www.scipopt.org/doc/html/EVENT.php

There are a few events which will get triggered when a primal solution is found (SCIP_EVENTTYPE_POORSOLFOUND, SCIP_EVENTTYPE_BESTSOLFOUND):
https://www.scipopt.org/doc/html/type__event_8h.php

The Pyscipopt API is pretty straightforward. See here:
https://github.com/scipopt/PySCIPOpt/blob/master/tests/test_event.py

**Reply by cwfparsonson** [reply]
Thank you!
